proof_ipa.go

```diff
@@ -414,25 +422,57 @@ func PreStateTreeFromProof(proof *Proof, rootC *Point) (VerkleNode, error) { //
 		}
 		switch si.stemType {
 		case extStatusAbsentEmpty:
+			// All keys that are part of a proof of absence, must contain empty
+			// pre and post state values. If that isn't the case, the proof is invalid.
+			for i, k := range proof.Keys { // TODO: DoS risk, use map or binary search.
+				if bytes.HasPrefix(k, path) {
+					if proof.PreValues[i] != nil {
+						return nil, fmt.Errorf("proof of absence (empty) stem %x has a value", si.stem)
+					}
+				}
+			}
 		case extStatusAbsentOther:
 			si.stem = poas[0]
 			poas = poas[1:]
+			// All keys that are part of a proof of absence, must contain empty
+			// pre and post state values. If that isn't the case, the proof is invalid.
+			for i, k := range proof.Keys { // TODO: DoS risk, use map or binary search.
+				if bytes.HasPrefix(k, si.stem) {
+					if proof.PreValues[i] != nil {
+						return nil, fmt.Errorf("proof of absence (other) stem %x has a value", si.stem)
+					}
+				}
+			}
 		default:
 			// the first stem could be missing (e.g. the second stem in the
 			// group is the one that is present. Compare each key to the first
 			// stem, along the length of the path only.
 			stemPath := stems[stemIndex][:len(path)]
 			si.values = map[byte][]byte{}
-			for i, k := range proof.Keys {
+			for i, k := range proof.Keys { // TODO: DoS risk, use map or binary search.
 				if bytes.Equal(k[:len(path)], stemPath) && proof.PreValues[i] != nil {
 					si.values[k[31]] = proof.PreValues[i]
 					si.has_c1 = si.has_c1 || (k[31] < 128)
 					si.has_c2 = si.has_c2 || (k[31] >= 128)
 					// This key has values, its stem is the one that
 					// is present.
-					si.stem = k[:31]
+					if si.stem == nil {
+						si.stem = k[:31]
+						continue
+					}
+					// Any other key with values must have the same
+					// same previously detected stem. If that isn't the case,
+					// the proof is invalid.
+					if !bytes.Equal(si.stem, k[:31]) {
+						return nil, fmt.Errorf("multiple keys with values found for stem %x", k[:31])
+					}
```

Comment:
OK, this was me being paranoid but it's an important check.

Before in our code, we always did `si.stem = k[:31]`. Technically, if there were multiple keys in the same stem, it will reassign to the same value. Totally fine since that's not very relevant.

Now, this extra check in L466 is relevant. If we find some key that apparently matches the stem-path that we're processing but the full stem is different, then something hairy is happening. 

Somebody is trying to inject a claim about a key that might go in this place in the tree, but doesn't correspond to this Leaf. In normal circumstances this shouldn't happen, since if that was the case, in this place in the tree we wouldn't have a LeafNode but an Internal node that would distinguish both paths. But if I manually craft a proof dishonestly, I can create this situation that should be rejected. (Note that if we're at this place, this injected key has a value non-nil in `Values`; so clearly something is really bad (i.e: this isn't a proof of absence case))